### PR TITLE
Discotheque disco collection

### PR DESCRIPTION
**Pull request by logarithm**

### CODE REVIEW

**Comment by logarithm** [comment]
@drcrallen added

File: src/main/scala/com/metamx/common/scala/net/curator/Discotheque.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    discoZkConnect: String,
+    discoPath: String,
+    discoAnnounceConfig: Option[DiscoAnnounceConfig] = None,
+    discoZkTimeout: Duration = Duration.parse("PT15S")
+  ): Disco = {
+    val endpoint = DiscoEndpoint(discoZkConnect, discoPath)
+
+    val (disco, announceConfig) = discos.getOrElseUpdate(endpoint, {
+      val curatorConfig = endpoint.toCuratorConfig(discoZkTimeout)
+      val curator = curators.getOrElseUpdate(curatorConfig, Curator.create(curatorConfig, lifecycle))
```
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/net/curator/Discotheque.scala]: `Curator.create(curatorConfig, lifecycle)` is not a factory, right? so is it creating new curators and registering them in the lifecycle every call to `disco`?

Comment (logarithm) [inline src/main/scala/com/metamx/common/scala/net/curator/Discotheque.scala]: Second parameter of `getOrElseUpdate` is lazy function, so new curator will be created only of we don't have it in `curators` map

Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/net/curator/Discotheque.scala]: Gotcha

```scala
def getOrElseUpdate(key: K, op: => V): V = {
```

**Pull request merged**